// cojson/dist/coValues/account.js
import { logger } from "../logger.js";
import { RawCoMap } from "./coMap.js";
import { RawGroup } from "./group.js";
export function accountHeaderForInitialAgentSecret(agentSecret, crypto) {
    const agent = crypto.getAgentID(agentSecret);
    return {
        type: "comap",
        ruleset: { type: "group", initialAdmin: agent },
        meta: {
            type: "account",
        },
        createdAt: null,
        uniqueness: null,
    };
}
export class RawAccount extends RawGroup {
    currentAgentID() {
        if (this._cachedCurrentAgentID) {
            return this._cachedCurrentAgentID;
        }
        const agents = this.keys()
            .filter((k) => k.startsWith("sealer_"))
            .sort((a, b) => (this.lastEditAt(a)?.at.getTime() || 0) -
            (this.lastEditAt(b)?.at.getTime() || 0));
        if (agents.length !== 1) {
            logger.warn("Account has " + agents.length + " agents", { id: this.id });
        }
        this._cachedCurrentAgentID = agents[0];
        return agents[0];
    }
    createInvite(_) {
        throw new Error("Cannot create invite from an account");
    }
}
/** @hidden */
export class ControlledAccount {
    constructor(account, agentSecret) {
        this.account = account;
        this.agentSecret = agentSecret;
        this.crypto = account.core.node.crypto;
    }
    get id() {
        return this.account.id;
    }
    currentAgentID() {
        if (this._cachedCurrentAgentID) {
            return this._cachedCurrentAgentID;
        }
        const agentID = this.crypto.getAgentID(this.agentSecret);
        this._cachedCurrentAgentID = agentID;
        return agentID;
    }
    currentSignerID() {
        return this.crypto.getAgentSignerID(this.currentAgentID());
    }
    currentSignerSecret() {
        return this.crypto.getAgentSignerSecret(this.agentSecret);
    }
    currentSealerID() {
        return this.crypto.getAgentSealerID(this.currentAgentID());
    }
    currentSealerSecret() {
        return this.crypto.getAgentSealerSecret(this.agentSecret);
    }
}
export class ControlledAgent {
    constructor(agentSecret, crypto) {
        this.agentSecret = agentSecret;
        this.crypto = crypto;
    }
    get id() {
        return this.crypto.getAgentID(this.agentSecret);
    }
    currentAgentID() {
        return this.crypto.getAgentID(this.agentSecret);
    }
    currentSignerID() {
        return this.crypto.getAgentSignerID(this.currentAgentID());
    }
    currentSignerSecret() {
        return this.crypto.getAgentSignerSecret(this.agentSecret);
    }
    currentSealerID() {
        return this.crypto.getAgentSealerID(this.currentAgentID());
    }
    currentSealerSecret() {
        return this.crypto.getAgentSealerSecret(this.agentSecret);
    }
}
export class RawProfile extends RawCoMap {
}
export function expectAccount(content) {
    if (!(content instanceof RawAccount)) {
        throw new Error("Expected an account");
    }
    return content;
}
