// cojson-storage/dist/sqlite/client.js
import { logger } from "cojson";
export function getErrorMessage(error) {
    return error instanceof Error ? error.message : "Unknown error";
}
export class SQLiteClient {
    constructor(db) {
        this.db = db;
    }
    getCoValue(coValueId) {
        const coValueRow = this.db.get("SELECT * FROM coValues WHERE id = ?", [coValueId]);
        if (!coValueRow)
            return;
        try {
            const parsedHeader = (coValueRow?.header &&
                JSON.parse(coValueRow.header));
            return {
                ...coValueRow,
                header: parsedHeader,
            };
        }
        catch (e) {
            const headerValue = coValueRow?.header ?? "";
            logger.warn(`Invalid JSON in header: ${headerValue}`, {
                id: coValueId,
                err: e,
            });
            return;
        }
    }
    getCoValueSessions(coValueRowId) {
        return this.db.query("SELECT * FROM sessions WHERE coValue = ?", [coValueRowId]);
    }
    getSingleCoValueSession(coValueRowId, sessionID) {
        return this.db.get("SELECT * FROM sessions WHERE coValue = ? AND sessionID = ?", [coValueRowId, sessionID]);
    }
    getNewTransactionInSession(sessionRowId, fromIdx, toIdx) {
        const txs = this.db.query("SELECT * FROM transactions WHERE ses = ? AND idx >= ? AND idx <= ?", [sessionRowId, fromIdx, toIdx]);
        try {
            return txs.map((transactionRow) => ({
                ...transactionRow,
                tx: JSON.parse(transactionRow.tx),
            }));
        }
        catch (e) {
            logger.warn("Invalid JSON in transaction", { err: e });
            return [];
        }
    }
    getSignatures(sessionRowId, firstNewTxIdx) {
        return this.db.query("SELECT * FROM signatureAfter WHERE ses = ? AND idx >= ?", [sessionRowId, firstNewTxIdx]);
    }
    addCoValue(msg) {
        const result = this.db.get("INSERT INTO coValues (id, header) VALUES (?, ?) RETURNING rowID", [msg.id, JSON.stringify(msg.header)]);
        if (!result) {
            throw new Error("Failed to add coValue");
        }
        return result.rowID;
    }
    addSessionUpdate({ sessionUpdate, }) {
        const result = this.db.get(`INSERT INTO sessions (coValue, sessionID, lastIdx, lastSignature, bytesSinceLastSignature) VALUES (?, ?, ?, ?, ?)
                            ON CONFLICT(coValue, sessionID) DO UPDATE SET lastIdx=excluded.lastIdx, lastSignature=excluded.lastSignature, bytesSinceLastSignature=excluded.bytesSinceLastSignature
                            RETURNING rowID`, [
            sessionUpdate.coValue,
            sessionUpdate.sessionID,
            sessionUpdate.lastIdx,
            sessionUpdate.lastSignature,
            sessionUpdate.bytesSinceLastSignature,
        ]);
        if (!result) {
            throw new Error("Failed to add session update");
        }
        return result.rowID;
    }
    addTransaction(sessionRowID, nextIdx, newTransaction) {
        this.db.run("INSERT INTO transactions (ses, idx, tx) VALUES (?, ?, ?)", [
            sessionRowID,
            nextIdx,
            JSON.stringify(newTransaction),
        ]);
    }
    addSignatureAfter({ sessionRowID, idx, signature, }) {
        this.db.run("INSERT INTO signatureAfter (ses, idx, signature) VALUES (?, ?, ?)", [sessionRowID, idx, signature]);
    }
    transaction(operationsCallback) {
        this.db.transaction(operationsCallback);
        return undefined;
    }
}
