// cojson/dist/localNode.js
import { err, ok } from "neverthrow";
import { CO_VALUE_LOADING_CONFIG, CoValueCore, idforHeader, } from "./coValueCore/coValueCore.js";
import { VerifiedState, } from "./coValueCore/verifiedState.js";
import { ControlledAccount, ControlledAgent, accountHeaderForInitialAgentSecret, expectAccount, } from "./coValues/account.js";
import { secretSeedFromInviteSecret, } from "./coValues/group.js";
import { isAgentID } from "./ids.js";
import { logger } from "./logger.js";
import { SyncManager } from "./sync.js";
import { accountOrAgentIDfromSessionID } from "./typeUtils/accountOrAgentIDfromSessionID.js";
import { expectGroup } from "./typeUtils/expectGroup.js";
/** A `LocalNode` represents a local view of a set of loaded `CoValue`s, from the perspective of a particular account (or primitive cryptographic agent).

A `LocalNode` can have peers that it syncs to, for example some form of local persistence, or a sync server, such as `cloud.jazz.tools` (Jazz Cloud).

@example
You typically get hold of a `LocalNode` using `jazz-react`'s `useJazz()`:

```typescript
const { localNode } = useJazz();
```
*/
export class LocalNode {
    /** @category 3. Low-level */
    constructor(agentSecret, currentSessionID, crypto) {
        /** @internal */
        this.coValues = new Map();
        /** @category 3. Low-level */
        this.syncManager = new SyncManager(this);
        this.crashed = undefined;
        this.agentSecret = agentSecret;
        this.currentSessionID = currentSessionID;
        this.crypto = crypto;
    }
    getCoValue(id) {
        let entry = this.coValues.get(id);
        if (!entry) {
            entry = CoValueCore.fromID(id, this);
            this.coValues.set(id, entry);
        }
        return entry;
    }
    allCoValues() {
        return this.coValues.values();
    }
    putCoValue(id, verified, { forceOverwrite = false } = {}) {
        const entry = this.getCoValue(id);
        entry.internalMarkMagicallyAvailable(verified, { forceOverwrite });
        return entry;
    }
    internalDeleteCoValue(id) {
        this.coValues.delete(id);
    }
    getCurrentAccountOrAgentID() {
        return accountOrAgentIDfromSessionID(this.currentSessionID);
    }
    getCurrentAgent() {
        const accountOrAgent = this.getCurrentAccountOrAgentID();
        if (isAgentID(accountOrAgent)) {
            return new ControlledAgent(this.agentSecret, this.crypto);
        }
        return new ControlledAccount(expectAccount(this.expectCoValueLoaded(accountOrAgent).getCurrentContent()), this.agentSecret);
    }
    expectCurrentAccountID(reason) {
        const accountOrAgent = this.getCurrentAccountOrAgentID();
        if (isAgentID(accountOrAgent)) {
            throw new Error("Current account is an agent, but expected an account: " + reason);
        }
        return accountOrAgent;
    }
    expectCurrentAccount(reason) {
        const accountID = this.expectCurrentAccountID(reason);
        return expectAccount(this.expectCoValueLoaded(accountID).getCurrentContent());
    }
    static internalCreateAccount(opts) {
        const { crypto, initialAgentSecret = crypto.newRandomAgentSecret(), peersToLoadFrom = [], } = opts;
        const accountHeader = accountHeaderForInitialAgentSecret(initialAgentSecret, crypto);
        const accountID = idforHeader(accountHeader, crypto);
        const node = new LocalNode(initialAgentSecret, crypto.newRandomSessionID(accountID), crypto);
        for (const peer of peersToLoadFrom) {
            node.syncManager.addPeer(peer);
        }
        const accountAgentID = crypto.getAgentID(initialAgentSecret);
        const rawAccount = expectGroup(node.createCoValue(accountHeader).getCurrentContent());
        rawAccount.set(accountAgentID, "admin", "trusting");
        const readKey = crypto.newRandomKeySecret();
        const sealed = crypto.seal({
            message: readKey.secret,
            from: crypto.getAgentSealerSecret(initialAgentSecret),
            to: crypto.getAgentSealerID(accountAgentID),
            nOnceMaterial: {
                in: rawAccount.id,
                tx: rawAccount.core.nextTransactionID(),
            },
        });
        rawAccount.set(`${readKey.id}_for_${accountAgentID}`, sealed, "trusting");
        rawAccount.set("readKey", readKey.id, "trusting");
        return node.expectCurrentAccount("after creation");
    }
    /** @category 2. Node Creation */
    static async withNewlyCreatedAccount({ creationProps, peersToLoadFrom, migration, crypto, initialAgentSecret = crypto.newRandomAgentSecret(), }) {
        const account = LocalNode.internalCreateAccount({
            crypto,
            initialAgentSecret,
            peersToLoadFrom,
        });
        const node = account.core.node;
        if (migration) {
            await migration(account, node, creationProps);
        }
        else {
            const profileGroup = node.createGroup();
            profileGroup.addMember("everyone", "reader");
            const profile = profileGroup.createMap({
                name: creationProps.name,
            });
            account.set("profile", profile.id, "trusting");
        }
        const profileId = account.get("profile");
        if (!profileId) {
            throw new Error("Must set account profile in initial migration");
        }
        if (node.syncManager.hasStoragePeers()) {
            await Promise.all([
                node.syncManager.waitForStorageSync(account.id),
                node.syncManager.waitForStorageSync(profileId),
            ]);
        }
        return {
            node,
            accountID: account.id,
            accountSecret: initialAgentSecret,
            sessionID: node.currentSessionID,
        };
    }
    /** @category 2. Node Creation */
    static async withLoadedAccount({ accountID, accountSecret, sessionID, peersToLoadFrom, crypto, migration, }) {
        try {
            const node = new LocalNode(accountSecret, sessionID || crypto.newRandomSessionID(accountID), crypto);
            for (const peer of peersToLoadFrom) {
                node.syncManager.addPeer(peer);
            }
            const account = await node.load(accountID);
            if (account === "unavailable") {
                throw new Error("Account unavailable from all peers");
            }
            const profileID = account.get("profile");
            if (!profileID) {
                throw new Error("Account has no profile");
            }
            // Preload the profile
            await node.load(profileID);
            if (migration) {
                await migration(account, node);
            }
            return node;
        }
        catch (e) {
            logger.error("Error withLoadedAccount", { err: e });
            throw e;
        }
    }
    /** @internal */
    createCoValue(header) {
        if (this.crashed) {
            throw new Error("Trying to create CoValue after node has crashed", {
                cause: this.crashed,
            });
        }
        const id = idforHeader(header, this.crypto);
        const coValue = this.putCoValue(id, new VerifiedState(id, this.crypto, header, new Map()));
        void this.syncManager.requestCoValueSync(coValue);
        return coValue;
    }
    /** @internal */
    async loadCoValueCore(id, skipLoadingFromPeer) {
        if (!id) {
            throw new Error("Trying to load CoValue with undefined id");
        }
        if (!id.startsWith("co_z")) {
            throw new Error(`Trying to load CoValue with invalid id ${id}`);
        }
        if (this.crashed) {
            throw new Error("Trying to load CoValue after node has crashed", {
                cause: this.crashed,
            });
        }
        let retries = 0;
        while (true) {
            const coValue = this.getCoValue(id);
            if (coValue.isAvailable()) {
                return coValue;
            }
            if (coValue.loadingState === "unknown" ||
                coValue.loadingState === "unavailable") {
                const peers = this.syncManager.getServerAndStoragePeers(skipLoadingFromPeer);
                if (peers.length === 0) {
                    return coValue;
                }
                coValue.loadFromPeers(peers).catch((e) => {
                    logger.error("Error loading from peers", {
                        id,
                        err: e,
                    });
                });
            }
            const result = await coValue.waitForAvailableOrUnavailable();
            if (result.isAvailable() ||
                retries >= CO_VALUE_LOADING_CONFIG.MAX_RETRIES) {
                return result;
            }
            await Promise.race([
                new Promise((resolve) => setTimeout(resolve, CO_VALUE_LOADING_CONFIG.RETRY_DELAY)),
                coValue.waitForAvailable(), // Stop waiting if the coValue becomes available
            ]);
            retries++;
        }
    }
    /**
     * Loads a CoValue's content, syncing from peers as necessary and resolving the returned
     * promise once a first version has been loaded. See `coValue.subscribe()` and `node.useTelepathicData()`
     * for listening to subsequent updates to the CoValue.
     *
     * @category 3. Low-level
     */
    async load(id) {
        const core = await this.loadCoValueCore(id);
        if (!core.isAvailable()) {
            return "unavailable";
        }
        return core.getCurrentContent();
    }
    getLoaded(id) {
        const coValue = this.getCoValue(id);
        if (coValue.isAvailable()) {
            return coValue.getCurrentContent();
        }
        return undefined;
    }
    /** @category 3. Low-level */
    subscribe(id, callback) {
        let stopped = false;
        let unsubscribe;
        this.load(id)
            .then((coValue) => {
            if (stopped) {
                return;
            }
            if (coValue === "unavailable") {
                callback("unavailable");
                return;
            }
            unsubscribe = coValue.subscribe(callback);
        })
            .catch((e) => {
            logger.error("Subscription error", {
                id,
                err: e,
            });
        });
        return () => {
            stopped = true;
            unsubscribe?.();
        };
    }
    async acceptInvite(groupOrOwnedValueID, inviteSecret) {
        const value = await this.load(groupOrOwnedValueID);
        if (value === "unavailable") {
            throw new Error("Trying to accept invite: Group/owned value unavailable from all peers");
        }
        const ruleset = value.core.verified.header.ruleset;
        let group;
        if (ruleset.type === "unsafeAllowAll") {
            throw new Error("Can only accept invites to values owned by groups");
        }
        if (ruleset.type === "ownedByGroup") {
            const owner = await this.load(ruleset.group);
            if (owner === "unavailable") {
                throw new Error("Trying to accept invite: CoValue owner unavailable from all peers");
            }
            group = expectGroup(owner);
        }
        else {
            group = expectGroup(value);
        }
        if (group.core.verified.header.meta?.type === "account") {
            throw new Error("Can't accept invites to values owned by accounts");
        }
        const inviteAgentSecret = this.crypto.agentSecretFromSecretSeed(secretSeedFromInviteSecret(inviteSecret));
        const inviteAgentID = this.crypto.getAgentID(inviteAgentSecret);
        const inviteRole = await new Promise((resolve, reject) => {
            group.subscribe((groupUpdate) => {
                const role = groupUpdate.get(inviteAgentID);
                if (role) {
                    resolve(role);
                }
            });
            setTimeout(() => reject(new Error("Couldn't find invite before timeout")), 2000);
        });
        if (!inviteRole) {
            throw new Error("No invite found");
        }
        const account = this.getCurrentAgent();
        const existingRole = group.get(account.id);
        if (existingRole === "admin" ||
            (existingRole === "writer" && inviteRole === "writerInvite") ||
            (existingRole === "writer" && inviteRole === "reader") ||
            (existingRole === "reader" && inviteRole === "readerInvite") ||
            (existingRole && inviteRole === "writeOnlyInvite")) {
            logger.debug("Not accepting invite that would replace or downgrade role");
            return;
        }
        const groupAsInvite = expectGroup(this.loadCoValueAsDifferentAgent(group.id, inviteAgentSecret).getCurrentContent());
        groupAsInvite.addMemberInternal(account, inviteRole === "adminInvite"
            ? "admin"
            : inviteRole === "writerInvite"
                ? "writer"
                : inviteRole === "writeOnlyInvite"
                    ? "writeOnly"
                    : "reader");
        group.core.internalShamefullyCloneVerifiedStateFrom(groupAsInvite.core.verified, { forceOverwrite: true });
        group.processNewTransactions();
        group.core.notifyUpdate("immediate");
        this.syncManager.requestCoValueSync(group.core);
    }
    /** @internal */
    expectCoValueLoaded(id, expectation) {
        const coValue = this.getCoValue(id);
        if (!coValue.isAvailable()) {
            throw new Error(`${expectation ? expectation + ": " : ""}CoValue ${id} not yet loaded.`);
        }
        return coValue;
    }
    /** @internal */
    expectProfileLoaded(id, expectation) {
        const account = this.expectCoValueLoaded(id, expectation);
        const profileID = expectGroup(account.getCurrentContent()).get("profile");
        if (!profileID) {
            throw new Error(`${expectation ? expectation + ": " : ""}Account ${id} has no profile`);
        }
        return this.expectCoValueLoaded(profileID, expectation).getCurrentContent();
    }
    /** @internal */
    resolveAccountAgent(id, expectation) {
        if (isAgentID(id)) {
            return ok(id);
        }
        let coValue;
        try {
            coValue = this.expectCoValueLoaded(id, expectation);
        }
        catch (e) {
            return err({
                type: "ErrorLoadingCoValueCore",
                expectation,
                id,
                error: e,
            });
        }
        if (coValue.verified.header.type !== "comap" ||
            coValue.verified.header.ruleset.type !== "group" ||
            !coValue.verified.header.meta ||
            !("type" in coValue.verified.header.meta) ||
            coValue.verified.header.meta.type !== "account") {
            return err({
                type: "UnexpectedlyNotAccount",
                expectation,
                id,
            });
        }
        return ok(coValue.getCurrentContent().currentAgentID());
    }
    createGroup(uniqueness = this.crypto.createdNowUnique()) {
        const account = this.getCurrentAgent();
        const groupCoValue = this.createCoValue({
            type: "comap",
            ruleset: { type: "group", initialAdmin: account.id },
            meta: null,
            ...uniqueness,
        });
        const group = expectGroup(groupCoValue.getCurrentContent());
        group.set(account.id, "admin", "trusting");
        const readKey = this.crypto.newRandomKeySecret();
        group.set(`${readKey.id}_for_${account.id}`, this.crypto.seal({
            message: readKey.secret,
            from: account.currentSealerSecret(),
            to: account.currentSealerID(),
            nOnceMaterial: {
                in: groupCoValue.id,
                tx: groupCoValue.nextTransactionID(),
            },
        }), "trusting");
        group.set("readKey", readKey.id, "trusting");
        return group;
    }
    loadCoValueAsDifferentAgent(id, secret, accountId) {
        const agent = new ControlledAgent(secret, this.crypto);
        const newNode = new LocalNode(secret, this.crypto.newRandomSessionID(accountId || agent.id), this.crypto);
        newNode.cloneVerifiedStateFrom(this, id);
        return newNode.expectCoValueLoaded(id);
    }
    /** @internal */
    cloneVerifiedStateFrom(otherNode, id) {
        const coValuesIdsToCopy = [id];
        // Scan all the dependencies and add them to the list
        for (let i = 0; i < coValuesIdsToCopy.length; i++) {
            const coValueID = coValuesIdsToCopy[i];
            const coValue = otherNode.getCoValue(coValueID);
            if (!coValue.isAvailable()) {
                continue;
            }
            for (const dep of coValue.getDependedOnCoValues()) {
                coValuesIdsToCopy.push(dep);
            }
        }
        // Copy the coValue all the dependencies by following the dependency order
        while (coValuesIdsToCopy.length > 0) {
            const coValueID = coValuesIdsToCopy.pop();
            const coValue = otherNode.getCoValue(coValueID);
            if (!coValue.isAvailable()) {
                continue;
            }
            if (this.coValues.get(coValueID)?.isAvailable()) {
                continue;
            }
            this.putCoValue(coValueID, coValue.verified);
        }
    }
    gracefulShutdown() {
        this.syncManager.gracefulShutdown();
    }
}
