// cojson/dist/sync.js
import { ValueType, metrics } from "@opentelemetry/api";
import { PeerState } from "./PeerState.js";
import { SyncStateManager } from "./SyncStateManager.js";
import { getDependedOnCoValuesFromRawData } from "./coValueCore/utils.js";
import { logger } from "./logger.js";
import { accountOrAgentIDfromSessionID } from "./typeUtils/accountOrAgentIDfromSessionID.js";
import { isAccountID } from "./typeUtils/isAccountID.js";
export function emptyKnownState(id) {
    return {
        id,
        header: false,
        sessions: {},
    };
}
export function combinedKnownStates(stateA, stateB) {
    const sessionStates = {};
    const allSessions = new Set([
        ...Object.keys(stateA.sessions),
        ...Object.keys(stateB.sessions),
    ]);
    for (const sessionID of allSessions) {
        const stateAValue = stateA.sessions[sessionID];
        const stateBValue = stateB.sessions[sessionID];
        sessionStates[sessionID] = Math.max(stateAValue || 0, stateBValue || 0);
    }
    return {
        id: stateA.id,
        header: stateA.header || stateB.header,
        sessions: sessionStates,
    };
}
export class SyncManager {
    constructor(local) {
        this.peers = {};
        this.peersCounter = metrics.getMeter("cojson").createUpDownCounter("jazz.peers", {
            description: "Amount of connected peers",
            valueType: ValueType.INT,
            unit: "peer",
        });
        this.requestedSyncs = new Set();
        this.local = local;
        this.syncState = new SyncStateManager(this);
        this.transactionsSizeHistogram = metrics
            .getMeter("cojson")
            .createHistogram("jazz.transactions.size", {
            description: "The size of transactions in a covalue",
            unit: "bytes",
            valueType: ValueType.INT,
        });
    }
    peersInPriorityOrder() {
        return Object.values(this.peers).sort((a, b) => {
            const aPriority = a.priority || 0;
            const bPriority = b.priority || 0;
            return bPriority - aPriority;
        });
    }
    getPeers() {
        return Object.values(this.peers);
    }
    getServerAndStoragePeers(excludePeerId) {
        return this.peersInPriorityOrder().filter((peer) => peer.isServerOrStoragePeer() &&
            peer.id !== excludePeerId &&
            !peer.closed);
    }
    hasStoragePeers() {
        return this.getPeers().some((peer) => peer.role === "storage" && !peer.closed);
    }
    handleSyncMessage(msg, peer) {
        if (msg.id === undefined || msg.id === null) {
            logger.warn("Received sync message with undefined id", {
                msg,
            });
            return;
        }
        else if (!msg.id.startsWith("co_z")) {
            logger.warn("Received sync message with invalid id", {
                msg,
            });
            return;
        }
        else if (this.local.getCoValue(msg.id).isErroredInPeer(peer.id)) {
            logger.warn(`Skipping message ${msg.action} on errored coValue ${msg.id} from peer ${peer.id}`);
            return;
        }
        // TODO: validate
        switch (msg.action) {
            case "load":
                return this.handleLoad(msg, peer);
            case "known":
                if (msg.isCorrection) {
                    return this.handleCorrection(msg, peer);
                }
                else {
                    return this.handleKnownState(msg, peer);
                }
            case "content":
                return this.handleNewContent(msg, peer);
            case "done":
                return this.handleUnsubscribe(msg);
            default:
                throw new Error(`Unknown message type ${msg.action}`);
        }
    }
    sendNewContentIncludingDependencies(id, peer) {
        const coValue = this.local.getCoValue(id);
        if (!coValue.isAvailable()) {
            return;
        }
        for (const dependency of coValue.getDependedOnCoValues()) {
            this.sendNewContentIncludingDependencies(dependency, peer);
        }
        const newContentPieces = coValue.verified.newContentSince(peer.optimisticKnownStates.get(id));
        if (newContentPieces) {
            for (const piece of newContentPieces) {
                this.trySendToPeer(peer, piece);
            }
            peer.combineOptimisticWith(id, coValue.knownState());
        }
        else if (!peer.toldKnownState.has(id)) {
            this.trySendToPeer(peer, {
                action: "known",
                ...coValue.knownState(),
            });
        }
        peer.trackToldKnownState(id);
    }
    startPeerReconciliation(peer) {
        const coValuesOrderedByDependency = [];
        const gathered = new Set();
        const buildOrderedCoValueList = (coValue) => {
            if (gathered.has(coValue.id)) {
                return;
            }
            gathered.add(coValue.id);
            for (const id of coValue.getDependedOnCoValues()) {
                const coValue = this.local.getCoValue(id);
                if (coValue.isAvailable()) {
                    buildOrderedCoValueList(coValue);
                }
            }
            coValuesOrderedByDependency.push(coValue);
        };
        for (const coValue of this.local.allCoValues()) {
            if (!coValue.isAvailable()) {
                // If the coValue is unavailable and we never tried this peer
                // we try to load it from the peer
                if (!peer.loadRequestSent.has(coValue.id)) {
                    peer.trackLoadRequestSent(coValue.id);
                    this.trySendToPeer(peer, {
                        action: "load",
                        header: false,
                        id: coValue.id,
                        sessions: {},
                    });
                }
            }
            else {
                // Build the list of coValues ordered by dependency
                // so we can send the load message in the correct order
                buildOrderedCoValueList(coValue);
            }
            // Fill the missing known states with empty known states
            if (!peer.optimisticKnownStates.has(coValue.id)) {
                peer.setOptimisticKnownState(coValue.id, "empty");
            }
        }
        for (const coValue of coValuesOrderedByDependency) {
            /**
             * We send the load messages to:
             * - Subscribe to the coValue updates
             * - Start the sync process in case we or the other peer
             *   lacks some transactions
             */
            peer.trackLoadRequestSent(coValue.id);
            this.trySendToPeer(peer, {
                action: "load",
                ...coValue.knownState(),
            });
        }
    }
    addPeer(peer) {
        const prevPeer = this.peers[peer.id];
        if (prevPeer && !prevPeer.closed) {
            prevPeer.gracefulShutdown();
        }
        const peerState = new PeerState(peer, prevPeer?.knownStates);
        this.peers[peer.id] = peerState;
        this.peersCounter.add(1, { role: peer.role });
        const unsubscribeFromKnownStatesUpdates = peerState.knownStates.subscribe((id) => {
            this.syncState.triggerUpdate(peer.id, id);
        });
        if (peerState.isServerOrStoragePeer()) {
            void this.startPeerReconciliation(peerState);
        }
        peerState
            .processIncomingMessages((msg) => {
            this.handleSyncMessage(msg, peerState);
        })
            .then(() => {
            if (peer.crashOnClose) {
                logger.error("Unexepcted close from peer", {
                    peerId: peer.id,
                    peerRole: peer.role,
                });
                this.local.crashed = new Error("Unexpected close from peer");
                throw new Error("Unexpected close from peer");
            }
        })
            .catch((e) => {
            logger.error("Error processing messages from peer", {
                err: e,
                peerId: peer.id,
                peerRole: peer.role,
            });
            if (peer.crashOnClose) {
                this.local.crashed = e;
                throw new Error(e);
            }
        })
            .finally(() => {
            peerState.gracefulShutdown();
            unsubscribeFromKnownStatesUpdates();
            this.peersCounter.add(-1, { role: peer.role });
            if (peer.deletePeerStateOnClose && this.peers[peer.id] === peerState) {
                delete this.peers[peer.id];
            }
        });
    }
    trySendToPeer(peer, msg) {
        return peer.pushOutgoingMessage(msg);
    }
    /**
     * Handles the load message from a peer.
     *
     * Differences with the known state message:
     * - The load message triggers the CoValue loading process on the other peer
     * - The peer known state is stored as-is instead of being merged
     * - The load message always replies with a known state message
     */
    handleLoad(msg, peer) {
        /**
         * We use the msg sessions as source of truth for the known states
         *
         * This way we can track part of the data loss that may occur when the other peer is restarted
         *
         */
        peer.setKnownState(msg.id, knownStateIn(msg));
        const coValue = this.local.getCoValue(msg.id);
        if (coValue.isAvailable()) {
            this.sendNewContentIncludingDependencies(msg.id, peer);
            return;
        }
        const eligiblePeers = this.getServerAndStoragePeers(peer.id);
        if (eligiblePeers.length === 0) {
            // We don't have any eligible peers to load the coValue from
            // so we send a known state back to the sender to let it know
            // that the coValue is unavailable
            peer.trackToldKnownState(msg.id);
            this.trySendToPeer(peer, {
                action: "known",
                id: msg.id,
                header: false,
                sessions: {},
            });
            return;
        }
        coValue.loadFromPeers(eligiblePeers).catch((e) => {
            logger.error("Error loading coValue in handleLoad", { err: e });
        });
        // We need to return from handleLoad immediately and wait for the CoValue to be loaded
        // in a new task, otherwise we might block further incoming content messages that would
        // resolve the CoValue as available. This can happen when we receive fresh
        // content from a client, but we are a server with our own upstream server(s)
        coValue
            .waitForAvailableOrUnavailable()
            .then((value) => {
            if (!value.isAvailable()) {
                peer.trackToldKnownState(msg.id);
                this.trySendToPeer(peer, {
                    action: "known",
                    id: msg.id,
                    header: false,
                    sessions: {},
                });
                return;
            }
            this.sendNewContentIncludingDependencies(msg.id, peer);
        })
            .catch((e) => {
            logger.error("Error loading coValue in handleLoad loading state", {
                err: e,
            });
        });
    }
    handleKnownState(msg, peer) {
        const coValue = this.local.getCoValue(msg.id);
        peer.combineWith(msg.id, knownStateIn(msg));
        // The header is a boolean value that tells us if the other peer do have information about the header.
        // If it's false in this point it means that the coValue is unavailable on the other peer.
        const availableOnPeer = peer.optimisticKnownStates.get(msg.id)?.header;
        if (!availableOnPeer) {
            coValue.markNotFoundInPeer(peer.id);
        }
        if (coValue.isAvailable()) {
            this.sendNewContentIncludingDependencies(msg.id, peer);
        }
    }
    recordTransactionsSize(newTransactions, source) {
        for (const tx of newTransactions) {
            const txLength = tx.privacy === "private"
                ? tx.encryptedChanges.length
                : tx.changes.length;
            this.transactionsSizeHistogram.record(txLength, {
                source,
            });
        }
    }
    handleNewContent(msg, peer) {
        const coValue = this.local.getCoValue(msg.id);
        if (!coValue.verified) {
            if (!msg.header) {
                this.trySendToPeer(peer, {
                    action: "known",
                    isCorrection: true,
                    id: msg.id,
                    header: false,
                    sessions: {},
                });
                return;
            }
            const sessionIDs = Object.keys(msg.new);
            const transactions = Object.values(msg.new).map((content) => content.newTransactions);
            for (const dependency of getDependedOnCoValuesFromRawData(msg.id, msg.header, sessionIDs, transactions)) {
                const dependencyCoValue = this.local.getCoValue(dependency);
                if (!dependencyCoValue.isAvailable()) {
                    coValue.markMissingDependency(dependency);
                    if (!dependencyCoValue.verified) {
                        const peers = this.getServerAndStoragePeers();
                        // if the peer that sent the content is a client, we add it to the list of peers
                        // to also ask them for the dependency
                        if (peer.role === "client") {
                            peers.push(peer);
                        }
                        dependencyCoValue.loadFromPeers(peers);
                    }
                }
            }
            peer.updateHeader(msg.id, true);
            coValue.provideHeader(msg.header, peer.id);
        }
        if (!coValue.verified) {
            throw new Error("Unreachable: CoValue should always have a verified state at this point");
        }
        let invalidStateAssumed = false;
        for (const [sessionID, newContentForSession] of Object.entries(msg.new)) {
            const ourKnownTxIdx = coValue.verified.sessions.get(sessionID)?.transactions.length;
            const theirFirstNewTxIdx = newContentForSession.after;
            if ((ourKnownTxIdx || 0) < theirFirstNewTxIdx) {
                invalidStateAssumed = true;
                continue;
            }
            const alreadyKnownOffset = ourKnownTxIdx
                ? ourKnownTxIdx - theirFirstNewTxIdx
                : 0;
            const newTransactions = newContentForSession.newTransactions.slice(alreadyKnownOffset);
            if (newTransactions.length === 0) {
                continue;
            }
            const accountId = accountOrAgentIDfromSessionID(sessionID);
            if (isAccountID(accountId)) {
                const account = this.local.getCoValue(accountId);
                // We can't verify the transaction without the account, so we delay the session content handling until the account is available
                if (!account.isAvailable()) {
                    // This covers the case where we are getting a new session on an already loaded coValue
                    // where we need to load the account to get their public key
                    if (!coValue.missingDependencies.has(accountId)) {
                        const peers = this.getServerAndStoragePeers();
                        if (peer.role === "client") {
                            // if the peer that sent the content is a client, we add it to the list of peers
                            // to also ask them for the dependency
                            peers.push(peer);
                        }
                        account.loadFromPeers(peers);
                    }
                    // We need to wait for the account to be available before we can verify the transaction
                    // Currently doing this by delaying the handleNewContent for the session to when we have the account
                    //
                    // This is not the best solution, because the knownState is not updated and the ACK response will be given
                    // by excluding the session.
                    // This is good enough implementation for now because the only case for the account to be missing are out-of-order
                    // dependencies push, so the gap should be short lived.
                    //
                    // When we are going to have sharded-peers we should revisit this, and store unverified sessions that are considered as part of the
                    // knwonState, but not actively used until they can be verified.
                    void account.waitForAvailable().then(() => {
                        this.handleNewContent({
                            action: "content",
                            id: coValue.id,
                            new: {
                                [sessionID]: newContentForSession,
                            },
                            priority: msg.priority,
                        }, peer);
                    });
                    continue;
                }
            }
            const result = coValue.tryAddTransactions(sessionID, newTransactions, undefined, newContentForSession.lastSignature, "immediate");
            if (result.isErr()) {
                console.error("Failed to add transactions", {
                    peerId: peer.id,
                    peerRole: peer.role,
                    id: msg.id,
                    err: result.error,
                });
                coValue.markErrored(peer.id, result.error);
                continue;
            }
            this.recordTransactionsSize(newTransactions, peer.role);
            peer.updateSessionCounter(msg.id, sessionID, newContentForSession.after +
                newContentForSession.newTransactions.length);
        }
        if (invalidStateAssumed) {
            this.trySendToPeer(peer, {
                action: "known",
                isCorrection: true,
                ...coValue.knownState(),
            });
            peer.trackToldKnownState(msg.id);
        }
        else {
            /**
             * We are sending a known state message to the peer to acknowledge the
             * receipt of the new content.
             *
             * This way the sender knows that the content has been received and applied
             * and can update their peer's knownState accordingly.
             */
            this.trySendToPeer(peer, {
                action: "known",
                ...coValue.knownState(),
            });
            peer.trackToldKnownState(msg.id);
        }
        const sourcePeer = peer;
        const syncedPeers = [];
        for (const peer of this.peersInPriorityOrder()) {
            /**
             * We sync the content against the source peer if it is a client or server peers
             * to upload any content that is available on the current node and not on the source peer.
             *
             * We don't need to do this with storage peers because we don't get updates from those peers,
             * only load and store content.
             */
            if (peer.id === sourcePeer.id && sourcePeer.role === "storage")
                continue;
            if (peer.closed)
                continue;
            if (coValue.isErroredInPeer(peer.id))
                continue;
            // We directly forward the new content to peers that have an active subscription
            if (peer.optimisticKnownStates.has(coValue.id)) {
                this.sendNewContentIncludingDependencies(coValue.id, peer);
                syncedPeers.push(peer);
            }
            else if (peer.isServerOrStoragePeer() &&
                !peer.loadRequestSent.has(coValue.id)) {
                const state = coValue.getStateForPeer(peer.id)?.type;
                // Check if there is a inflight load operation and we
                // are waiting for other peers to send the load request
                if (state === "unknown" || state === undefined) {
                    // Sending a load message to the peer to get to know how much content is missing
                    // before sending the new content
                    this.trySendToPeer(peer, {
                        action: "load",
                        ...coValue.knownState(),
                    });
                    peer.trackLoadRequestSent(coValue.id);
                    syncedPeers.push(peer);
                }
            }
        }
        for (const peer of syncedPeers) {
            this.syncState.triggerUpdate(peer.id, coValue.id);
        }
    }
    handleCorrection(msg, peer) {
        peer.setKnownState(msg.id, knownStateIn(msg));
        return this.sendNewContentIncludingDependencies(msg.id, peer);
    }
    handleUnsubscribe(_msg) { }
    requestCoValueSync(coValue) {
        if (this.requestedSyncs.has(coValue.id)) {
            return;
        }
        queueMicrotask(() => {
            if (this.requestedSyncs.has(coValue.id)) {
                this.syncCoValue(coValue);
            }
        });
        this.requestedSyncs.add(coValue.id);
    }
    syncCoValue(coValue) {
        this.requestedSyncs.delete(coValue.id);
        for (const peer of this.peersInPriorityOrder()) {
            if (peer.closed)
                continue;
            if (coValue.isErroredInPeer(peer.id))
                continue;
            // Only subscribed CoValues are synced to clients
            if (peer.role === "client" &&
                !peer.optimisticKnownStates.has(coValue.id)) {
                continue;
            }
            this.sendNewContentIncludingDependencies(coValue.id, peer);
        }
        for (const peer of this.getPeers()) {
            this.syncState.triggerUpdate(peer.id, coValue.id);
        }
    }
    waitForSyncWithPeer(peerId, id, timeout) {
        const { syncState } = this;
        const currentSyncState = syncState.getCurrentSyncState(peerId, id);
        const isTheConditionAlreadyMet = currentSyncState.uploaded;
        if (isTheConditionAlreadyMet) {
            return;
        }
        const peerState = this.peers[peerId];
        // The peer has been closed, so it isn't possible to sync
        if (!peerState || peerState.closed) {
            return;
        }
        // The client isn't subscribed to the coValue, so we won't sync it
        if (peerState.role === "client" &&
            !peerState.optimisticKnownStates.has(id)) {
            return;
        }
        return new Promise((resolve, reject) => {
            const unsubscribe = this.syncState.subscribeToPeerUpdates(peerId, (knownState, syncState) => {
                if (syncState.uploaded && knownState.id === id) {
                    resolve(true);
                    unsubscribe?.();
                    clearTimeout(timeoutId);
                }
            });
            const timeoutId = setTimeout(() => {
                reject(new Error(`Timeout waiting for sync on ${peerId}/${id}`));
                unsubscribe?.();
            }, timeout);
        });
    }
    waitForStorageSync(id, timeout = 30000) {
        const peers = this.getPeers();
        return Promise.all(peers
            .filter((peer) => peer.role === "storage")
            .map((peer) => this.waitForSyncWithPeer(peer.id, id, timeout)));
    }
    waitForSync(id, timeout = 30000) {
        const peers = this.getPeers();
        return Promise.all(peers.map((peer) => this.waitForSyncWithPeer(peer.id, id, timeout)));
    }
    waitForAllCoValuesSync(timeout = 60000) {
        const coValues = this.local.allCoValues();
        const validCoValues = Array.from(coValues).filter((coValue) => coValue.loadingState === "available" ||
            coValue.loadingState === "loading");
        return Promise.all(validCoValues.map((coValue) => this.waitForSync(coValue.id, timeout)));
    }
    gracefulShutdown() {
        for (const peer of Object.values(this.peers)) {
            peer.gracefulShutdown();
        }
    }
}
function knownStateIn(msg) {
    return {
        id: msg.id,
        header: msg.header,
        sessions: msg.sessions,
    };
}
